File: src/Read2Me.Core/Models/BookCommands.cs
namespace Read2Me.Core.Models;

public enum MergeDirection { Previous, Next }

/// <summary>
/// Where a new node lands relative to the anchor node it is inserted against.
/// Shared by every insertion command; the member names are also the JSON wire values.
/// </summary>
public enum InsertPosition { Before, After }

public abstract record BookCommand(ProjectFolderId FolderId);

// Delete
public record DeleteVolumeCommand(ProjectFolderId FolderId, Guid VolumeId) : BookCommand(FolderId);
public record DeletePartCommand(ProjectFolderId FolderId, Guid PartId) : BookCommand(FolderId);
public record DeleteChapterCommand(ProjectFolderId FolderId, Guid ChapterId) : BookCommand(FolderId);
public record DeleteParagraphCommand(ProjectFolderId FolderId, Guid ParagraphId) : BookCommand(FolderId);
public record DeleteParagraphItemCommand(ProjectFolderId FolderId, Guid ItemId) : BookCommand(FolderId);

// Update
public record UpdateVolumeTitleCommand(ProjectFolderId FolderId, Guid VolumeId, string Title) : BookCommand(FolderId);
public record UpdatePartTitleCommand(ProjectFolderId FolderId, Guid PartId, string Title) : BookCommand(FolderId);
public record UpdateChapterTitleCommand(ProjectFolderId FolderId, Guid ChapterId, string Title) : BookCommand(FolderId);
public record UpdateParagraphItemTextCommand(ProjectFolderId FolderId, Guid ItemId, string Text) : BookCommand(FolderId);

// Split — "split at X boundary to create new parent"
public record SplitAtPartCommand(ProjectFolderId FolderId, Guid PartId, string? NewVolumeTitle) : BookCommand(FolderId);
public record SplitAtChapterCommand(ProjectFolderId FolderId, Guid ChapterId, string? NewPartTitle) : BookCommand(FolderId);
public record SplitAtParagraphCommand(ProjectFolderId FolderId, Guid ParagraphId, string? NewChapterTitle) : BookCommand(FolderId);
public record SplitAtItemCommand(ProjectFolderId FolderId, Guid ItemId) : BookCommand(FolderId);

// Merge
public record MergeVolumeCommand(ProjectFolderId FolderId, Guid VolumeId, MergeDirection Direction) : BookCommand(FolderId);
public record MergePartCommand(ProjectFolderId FolderId, Guid PartId, MergeDirection Direction) : BookCommand(FolderId);
public record MergeChapterCommand(ProjectFolderId FolderId, Guid ChapterId, MergeDirection Direction) : BookCommand(FolderId);
public record MergeParagraphCommand(ProjectFolderId FolderId, Guid ParagraphId, MergeDirection Direction) : BookCommand(FolderId);
public record MergeParagraphItemCommand(ProjectFolderId FolderId, Guid ItemId, MergeDirection Direction) : BookCommand(FolderId);

// Character
public record SetItemCharacterCommand(ProjectFolderId FolderId, Guid ItemId, Guid? CharacterId) : BookCommand(FolderId);
public record CreateCharacterCommand(ProjectFolderId FolderId, string Name) : BookCommand(FolderId);
public record SetParagraphCharacterCommand(ProjectFolderId FolderId, Guid ParagraphId, Guid? CharacterId, string? VoiceInstructions = null) : BookCommand(FolderId);

/// <summary>
/// Stamps one speaker across every Character item in every listed paragraph — the bulk-assign
/// write, and the bulk sibling of <see cref="SetParagraphCharacterCommand"/>, which is left
/// unchanged and keeps its own callers.
/// </summary>
public record SetParagraphsCharacterCommand(ProjectFolderId FolderId, IReadOnlyList<Guid> ParagraphIds, Guid? CharacterId) : BookCommand(FolderId);

/// <summary>
/// One item's attribution, ready to apply: the item is addressed by id, and the speaker is already
/// resolved to a character id (null = unknown, which never erases an existing stamp).
/// <para>
/// Not to be confused with <c>AttributedItem</c>, the wire answer the LLM sends (index + speaker
/// name); this is the resolved apply-side record.
/// </para>
/// </summary>
public sealed record ItemAttribution(Guid ItemId, Guid? CharacterId, string? VoiceInstructions);

/// <summary>
/// Stamps speaker and voice instructions onto existing items of one paragraph. Item boundaries are
/// frozen (ADR 0005): this command never creates, deletes, reorders or retypes an item.
/// </summary>
public record AttributeItemsCommand(
    ProjectFolderId FolderId,
    Guid ParagraphId,
    IReadOnlyList<ItemAttribution> Items) : BookCommand(FolderId);

public record AddCharacterAliasCommand(ProjectFolderId FolderId, Guid CharacterId, string Name) : BookCommand(FolderId);
public record RemoveCharacterAliasCommand(ProjectFolderId FolderId, Guid AliasId) : BookCommand(FolderId);
public record MergeCharactersCommand(ProjectFolderId FolderId, Guid SurvivorId, Guid MergedId, bool AddNameAsAlias) : BookCommand(FolderId);
public record DeleteCharacterCommand(ProjectFolderId FolderId, Guid CharacterId) : BookCommand(FolderId);
public record RenameCharacterCommand(ProjectFolderId FolderId, Guid CharacterId, string Name) : BookCommand(FolderId);

// Narrator
/// <summary>
/// Points the book's narration at one of its Characters — Sherlock Holmes narrated by
/// Dr. Watson. <c>null</c> unlinks. The first project-scoped <see cref="BookCommand"/>:
/// every sibling addresses a node, character or voice.
/// </summary>
public record SetNarratorCharacterCommand(ProjectFolderId FolderId, Guid? CharacterId) : BookCommand(FolderId);

// Voice
public record CreateVoiceCommand(ProjectFolderId FolderId, Guid CharacterId, string Name, bool IsGenerated = false) : BookCommand(FolderId);
public record SetVoiceDefaultCommand(ProjectFolderId FolderId, Guid VoiceId) : BookCommand(FolderId);
public record UpdateVoiceCommand(ProjectFolderId FolderId, Guid VoiceId, string Name, string? Description) : BookCommand(FolderId);
public record SetVoiceDesignPromptCommand(ProjectFolderId FolderId, Guid VoiceId, string Prompt) : BookCommand(FolderId);
public record SetVoiceSettingsOverrideCommand(ProjectFolderId FolderId, Guid VoiceId, string? Json) : BookCommand(FolderId);
public record SetVoiceTtsSettingsOverrideCommand(ProjectFolderId FolderId, Guid VoiceId, string? Json) : BookCommand(FolderId);
public record SetVoiceTranscriptCommand(ProjectFolderId FolderId, Guid VoiceId, string Transcript) : BookCommand(FolderId);
public record SetVoiceAudioCommand(ProjectFolderId FolderId, Guid VoiceId, string AudioFileName) : BookCommand(FolderId);
public record SetVoiceGeneratedCommand(ProjectFolderId FolderId, Guid VoiceId, string AudioFileName, string Transcript, string DesignPrompt) : BookCommand(FolderId);
public record SetVoiceSourceCommand(ProjectFolderId FolderId, Guid VoiceId, bool IsGenerated) : BookCommand(FolderId);
public record DeleteVoiceCommand(ProjectFolderId FolderId, Guid VoiceId) : BookCommand(FolderId);

// Voice Rules
public enum RuleMoveDirection { Up, Down }
public record CreateVoiceRuleCommand(
    ProjectFolderId FolderId,
    Guid CharacterId,
    Guid VoiceId,
    VoiceAnchorLevel? FromLevel,
    Guid? FromNodeId,
    VoiceAnchorLevel? ToLevel,
    Guid? ToNodeId) : BookCommand(FolderId);
public record DeleteVoiceRuleCommand(ProjectFolderId FolderId, Guid RuleId) : BookCommand(FolderId);
public record MoveVoiceRuleCommand(ProjectFolderId FolderId, Guid RuleId, RuleMoveDirection Direction) : BookCommand(FolderId);

// Title insertion
public record AddBookTitleCommand(ProjectFolderId FolderId) : BookCommand(FolderId);
public record AddVolumeTitlesCommand(ProjectFolderId FolderId) : BookCommand(FolderId);
public record AddPartTitlesCommand(ProjectFolderId FolderId) : BookCommand(FolderId);
public record AddChapterTitlesCommand(ProjectFolderId FolderId) : BookCommand(FolderId);

// Item insertion
/// <summary>
/// Creates one Speech ParagraphItem next to <c>AnchorItemId</c>, inside the anchor's own Paragraph.
/// The new item is born unattributed — no speaker, no voice instructions, no audio: by construction
/// the anchor held two speakers, so inheriting its speaker would stamp a confident wrong answer that
/// looks attributed and never reaches the attribution queue.
/// </summary>
public record InsertParagraphItemCommand(ProjectFolderId FolderId, Guid AnchorItemId, InsertPosition Position, string Text) : BookCommand(FolderId);

// Pause insertion
public record AddPausesCommand(ProjectFolderId FolderId) : BookCommand(FolderId);
public record InsertPauseParagraphCommand(ProjectFolderId FolderId, Guid AnchorItemId, InsertPosition Position, PauseKind PauseKind) : BookCommand(FolderId);

public enum PauseKind { Pause, ParagraphPause, ChapterPause, PartPause, VolumePause }

// Clear
public record ClearBookContentCommand(ProjectFolderId FolderId) : BookCommand(FolderId);

// AI book edits
public enum BookEditTargetKind { VolumeTitle, PartTitle, ChapterTitle, ParagraphItemText }
public sealed record BookEditItem(BookEditTargetKind Kind, Guid Id, string NewValue);
public record ApplyBookEditsCommand(ProjectFolderId FolderId, IReadOnlyList<BookEditItem> Edits) : BookCommand(FolderId);

// Audio
public record SetParagraphItemAudioCommand(ProjectFolderId FolderId, Guid ItemId, string AudioFileName) : BookCommand(FolderId);

public enum AudioReviewState { NeedsReview, Dismissed }

public record SetAudioReviewCommand(
    ProjectFolderId FolderId,
    Guid ParagraphItemId,
    bool NormalizeOk,
    string? NormalizeReason,
    bool VerifyOk,
    double? Wer,
    string? VerifyReason,
    string? Transcript,
    string? OriginalTextSnapshot) : BookCommand(FolderId);

public record DismissAudioReviewCommand(ProjectFolderId FolderId, Guid ParagraphItemId) : BookCommand(FolderId);

File: src/Read2Me.Web/src/app/api/book-commands.ts
/**
 * Typed union of every command `POST /api/projects/{folder}/commands` accepts (ticket 05).
 *
 * Hand-written from `src/Read2Me.Core/Models/BookCommands.cs`: `type` is the record name minus the
 * `Command` suffix, properties are the record's positional parameters camelCased, `folderId` comes
 * from the route and is never sent. String enums (`MergeDirection`, `InsertPosition`, `PauseKind`,
 * `RuleMoveDirection`, `VoiceAnchorLevel`, `BookEditTargetKind`) travel as their member names.
 *
 * `BOOK_COMMAND_TYPES` is the runtime list; `book-commands.spec.ts` checks it against the host's
 * "Known types" 400 whenever a host is reachable, and the exhaustiveness check at the bottom fails
 * `npm run typecheck` when the union and the list drift apart.
 */
import type { Guid } from './dtos';

export type MergeDirection = 'Previous' | 'Next';
export type InsertPosition = 'Before' | 'After';
export type RuleMoveDirection = 'Up' | 'Down';
export type VoiceAnchorLevel = 'Volume' | 'Part' | 'Chapter' | 'Paragraph' | 'ParagraphItem';
export type BookEditTargetKind = 'VolumeTitle' | 'PartTitle' | 'ChapterTitle' | 'ParagraphItemText';
export type { PauseKind } from './dtos';
import type { PauseKind } from './dtos';

/** One item's resolved attribution (`ItemAttribution`); `characterId: null` never erases a stamp. */
export interface ItemAttribution {
  itemId: Guid;
  characterId: Guid | null;
  voiceInstructions?: string | null;
}

/** One AI book edit (`BookEditItem`). */
export interface BookEditItem {
  kind: BookEditTargetKind;
  id: Guid;
  newValue: string;
}

// ---- Delete ----------------------------------------------------------------------------------------
export interface DeleteVolume {
  type: 'DeleteVolume';
  volumeId: Guid;
}
export interface DeletePart {
  type: 'DeletePart';
  partId: Guid;
}
export interface DeleteChapter {
  type: 'DeleteChapter';
  chapterId: Guid;
}
export interface DeleteParagraph {
  type: 'DeleteParagraph';
  paragraphId: Guid;
}
export interface DeleteParagraphItem {
  type: 'DeleteParagraphItem';
  itemId: Guid;
}

// ---- Update ----------------------------------------------------------------------------------------
export interface UpdateVolumeTitle {
  type: 'UpdateVolumeTitle';
  volumeId: Guid;
  title: string;
}
export interface UpdatePartTitle {
  type: 'UpdatePartTitle';
  partId: Guid;
  title: string;
}
export interface UpdateChapterTitle {
  type: 'UpdateChapterTitle';
  chapterId: Guid;
  title: string;
}
export interface UpdateParagraphItemText {
  type: 'UpdateParagraphItemText';
  itemId: Guid;
  text: string;
}

// ---- Split -----------------------------------------------------------------------------------------
export interface SplitAtPart {
  type: 'SplitAtPart';
  partId: Guid;
  newVolumeTitle?: string | null;
}
export interface SplitAtChapter {
  type: 'SplitAtChapter';
  chapterId: Guid;
  newPartTitle?: string | null;
}
export interface SplitAtParagraph {
  type: 'SplitAtParagraph';
  paragraphId: Guid;
  newChapterTitle?: string | null;
}
export interface SplitAtItem {
  type: 'SplitAtItem';
  itemId: Guid;
}

// ---- Merge -----------------------------------------------------------------------------------------
export interface MergeVolume {
  type: 'MergeVolume';
  volumeId: Guid;
  direction: MergeDirection;
}
export interface MergePart {
  type: 'MergePart';
  partId: Guid;
  direction: MergeDirection;
}
export interface MergeChapter {
  type: 'MergeChapter';
  chapterId: Guid;
  direction: MergeDirection;
}
export interface MergeParagraph {
  type: 'MergeParagraph';
  paragraphId: Guid;
  direction: MergeDirection;
}
export interface MergeParagraphItem {
  type: 'MergeParagraphItem';
  itemId: Guid;
  direction: MergeDirection;
}

// ---- Characters ------------------------------------------------------------------------------------
export interface SetItemCharacter {
  type: 'SetItemCharacter';
  itemId: Guid;
  characterId: Guid | null;
}
export interface CreateCharacter {
  type: 'CreateCharacter';
  name: string;
}
export interface SetParagraphCharacter {
  type: 'SetParagraphCharacter';
  paragraphId: Guid;
  characterId: Guid | null;
  voiceInstructions?: string | null;
}
export interface SetParagraphsCharacter {
  type: 'SetParagraphsCharacter';
  paragraphIds: Guid[];
  characterId: Guid | null;
}
export interface AttributeItems {
  type: 'AttributeItems';
  paragraphId: Guid;
  items: ItemAttribution[];
}
export interface AddCharacterAlias {
  type: 'AddCharacterAlias';
  characterId: Guid;
  name: string;
}
export interface RemoveCharacterAlias {
  type: 'RemoveCharacterAlias';
  aliasId: Guid;
}
export interface MergeCharacters {
  type: 'MergeCharacters';
  survivorId: Guid;
  mergedId: Guid;
  addNameAsAlias: boolean;
}
export interface DeleteCharacter {
  type: 'DeleteCharacter';
  characterId: Guid;
}
export interface RenameCharacter {
  type: 'RenameCharacter';
  characterId: Guid;
  name: string;
}

// ---- Narrator --------------------------------------------------------------------------------------
/** `characterId: null` unlinks. 422 on an unknown character or a self-link. */
export interface SetNarratorCharacter {
  type: 'SetNarratorCharacter';
  characterId: Guid | null;
}

// ---- Voices ----------------------------------------------------------------------------------------
export interface CreateVoice {
  type: 'CreateVoice';
  characterId: Guid;
  name: string;
  isGenerated?: boolean;
}
export interface SetVoiceDefault {
  type: 'SetVoiceDefault';
  voiceId: Guid;
}
export interface UpdateVoice {
  type: 'UpdateVoice';
  voiceId: Guid;
  name: string;
  description?: string | null;
}
export interface SetVoiceDesignPrompt {
  type: 'SetVoiceDesignPrompt';
  voiceId: Guid;
  prompt: string;
}
export interface SetVoiceSettingsOverride {
  type: 'SetVoiceSettingsOverride';
  voiceId: Guid;
  json: string | null;
}
export interface SetVoiceTtsSettingsOverride {
  type: 'SetVoiceTtsSettingsOverride';
  voiceId: Guid;
  json: string | null;
}
export interface SetVoiceTranscript {
  type: 'SetVoiceTranscript';
  voiceId: Guid;
  transcript: string;
}
export interface SetVoiceAudio {
  type: 'SetVoiceAudio';
  voiceId: Guid;
  audioFileName: string;
}
export interface SetVoiceGenerated {
  type: 'SetVoiceGenerated';
  voiceId: Guid;
  audioFileName: string;
  transcript: string;
  designPrompt: string;
}
export interface SetVoiceSource {
  type: 'SetVoiceSource';
  voiceId: Guid;
  isGenerated: boolean;
}
export interface DeleteVoice {
  type: 'DeleteVoice';
  voiceId: Guid;
}

// ---- Voice rules -----------------------------------------------------------------------------------
export interface CreateVoiceRule {
  type: 'CreateVoiceRule';
  characterId: Guid;
  voiceId: Guid;
  fromLevel?: VoiceAnchorLevel | null;
  fromNodeId?: Guid | null;
  toLevel?: VoiceAnchorLevel | null;
  toNodeId?: Guid | null;
}
export interface DeleteVoiceRule {
  type: 'DeleteVoiceRule';
  ruleId: Guid;
}
export interface MoveVoiceRule {
  type: 'MoveVoiceRule';
  ruleId: Guid;
  direction: RuleMoveDirection;
}

// ---- Titles, pauses, insertion ---------------------------------------------------------------------
export interface AddBookTitle {
  type: 'AddBookTitle';
}
export interface AddVolumeTitles {
  type: 'AddVolumeTitles';
}
export interface AddPartTitles {
  type: 'AddPartTitles';
}
export interface AddChapterTitles {
  type: 'AddChapterTitles';
}
export interface AddPauses {
  type: 'AddPauses';
}
export interface InsertParagraphItem {
  type: 'InsertParagraphItem';
  anchorItemId: Guid;
  position: InsertPosition;
  text: string;
}
export interface InsertPauseParagraph {
  type: 'InsertPauseParagraph';
  anchorItemId: Guid;
  position: InsertPosition;
  pauseKind: PauseKind;
}

// ---- Other -----------------------------------------------------------------------------------------
export interface ClearBookContent {
  type: 'ClearBookContent';
}
export interface ApplyBookEdits {
  type: 'ApplyBookEdits';
  edits: BookEditItem[];
}
export interface SetParagraphItemAudio {
  type: 'SetParagraphItemAudio';
  itemId: Guid;
  audioFileName: string;
}
export interface SetAudioReview {
  type: 'SetAudioReview';
  paragraphItemId: Guid;
  normalizeOk: boolean;
  normalizeReason?: string | null;
  verifyOk: boolean;
  wer?: number | null;
  verifyReason?: string | null;
  transcript?: string | null;
  originalTextSnapshot?: string | null;
}
export interface DismissAudioReview {
  type: 'DismissAudioReview';
  paragraphItemId: Guid;
}

export type BookCommand =
  | DeleteVolume
  | DeletePart
  | DeleteChapter
  | DeleteParagraph
  | DeleteParagraphItem
  | UpdateVolumeTitle
  | UpdatePartTitle
  | UpdateChapterTitle
  | UpdateParagraphItemText
  | SplitAtPart
  | SplitAtChapter
  | SplitAtParagraph
  | SplitAtItem
  | MergeVolume
  | MergePart
  | MergeChapter
  | MergeParagraph
  | MergeParagraphItem
  | SetItemCharacter
  | CreateCharacter
  | SetParagraphCharacter
  | SetParagraphsCharacter
  | AttributeItems
  | AddCharacterAlias
  | RemoveCharacterAlias
  | MergeCharacters
  | DeleteCharacter
  | RenameCharacter
  | SetNarratorCharacter
  | CreateVoice
  | SetVoiceDefault
  | UpdateVoice
  | SetVoiceDesignPrompt
  | SetVoiceSettingsOverride
  | SetVoiceTtsSettingsOverride
  | SetVoiceTranscript
  | SetVoiceAudio
  | SetVoiceGenerated
  | SetVoiceSource
  | DeleteVoice
  | CreateVoiceRule
  | DeleteVoiceRule
  | MoveVoiceRule
  | AddBookTitle
  | AddVolumeTitles
  | AddPartTitles
  | AddChapterTitles
  | AddPauses
  | InsertParagraphItem
  | InsertPauseParagraph
  | ClearBookContent
  | ApplyBookEdits
  | SetParagraphItemAudio
  | SetAudioReview
  | DismissAudioReview;

export type BookCommandType = BookCommand['type'];

/** Every discriminator the host knows, alphabetical like its "Known types" 400 message. */
export const BOOK_COMMAND_TYPES = [
  'AddBookTitle',
  'AddChapterTitles',
  'AddCharacterAlias',
  'AddPartTitles',
  'AddPauses',
  'AddVolumeTitles',
  'ApplyBookEdits',
  'AttributeItems',
  'ClearBookContent',
  'CreateCharacter',
  'CreateVoice',
  'CreateVoiceRule',
  'DeleteChapter',
  'DeleteCharacter',
  'DeleteParagraph',
  'DeleteParagraphItem',
  'DeletePart',
  'DeleteVoice',
  'DeleteVoiceRule',
  'DeleteVolume',
  'DismissAudioReview',
  'InsertParagraphItem',
  'InsertPauseParagraph',
  'MergeChapter',
  'MergeCharacters',
  'MergeParagraph',
  'MergeParagraphItem',
  'MergePart',
  'MergeVolume',
  'MoveVoiceRule',
  'RemoveCharacterAlias',
  'RenameCharacter',
  'SetAudioReview',
  'SetItemCharacter',
  'SetNarratorCharacter',
  'SetParagraphCharacter',
  'SetParagraphItemAudio',
  'SetParagraphsCharacter',
  'SetVoiceAudio',
  'SetVoiceDefault',
  'SetVoiceDesignPrompt',
  'SetVoiceGenerated',
  'SetVoiceSettingsOverride',
  'SetVoiceSource',
  'SetVoiceTranscript',
  'SetVoiceTtsSettingsOverride',
  'SplitAtChapter',
  'SplitAtItem',
  'SplitAtParagraph',
  'SplitAtPart',
  'UpdateChapterTitle',
  'UpdateParagraphItemText',
  'UpdatePartTitle',
  'UpdateVoice',
  'UpdateVolumeTitle',
] as const satisfies readonly BookCommandType[];

/** Compile-time exhaustiveness: a union member missing from the list makes this `never`-typed. */
type MissingFromList = Exclude<BookCommandType, (typeof BOOK_COMMAND_TYPES)[number]>;
export const BOOK_COMMAND_LIST_COMPLETE: MissingFromList extends never ? true : never = true;
